core/remote_state_verifier.go

```diff
@@ -0,0 +1,438 @@
+				return nil, fmt.Errorf("block is nil, number: %d", number)
+			}
+		}
+	}
+
+	vm := &remoteVerifyManager{
+		bc:            blockchain,
+		tasks:         make(map[common.Hash]*verifyTask),
+		peers:         peers,
+		verifiedCache: verifiedCache,
+		allowInsecure: allowInsecure,
+
+		chainBlockCh: make(chan ChainHeadEvent, chainHeadChanSize),
+		verifyCh:     make(chan common.Hash, maxForkHeight),
+		messageCh:    make(chan verifyMessage),
+	}
+	vm.chainHeadSub = blockchain.SubscribeChainBlockEvent(vm.chainBlockCh)
+	return vm, nil
+}
+
+func (vm *remoteVerifyManager) mainLoop() {
+	defer vm.chainHeadSub.Unsubscribe()
+
+	pruneTicker := time.NewTicker(pruneInterval)
+	defer pruneTicker.Stop()
+	for {
+		select {
+		case h := <-vm.chainBlockCh:
+			vm.NewBlockVerifyTask(h.Block.Header())
+		case hash := <-vm.verifyCh:
+			vm.cacheBlockVerified(hash)
+			vm.taskLock.Lock()
+			if task, ok := vm.tasks[hash]; ok {
+				delete(vm.tasks, hash)
+				verifyTaskCounter.Dec(1)
+				verifyTaskSucceedMeter.Mark(1)
+				verifyTaskExecutionTimer.Update(time.Since(task.startAt))
+				task.Close()
+			}
+			vm.taskLock.Unlock()
+		case <-pruneTicker.C:
+			vm.taskLock.Lock()
+			for hash, task := range vm.tasks {
+				if vm.bc.insertStopped() || (vm.bc.CurrentHeader().Number.Cmp(task.blockHeader.Number) == 1 &&
+					vm.bc.CurrentHeader().Number.Uint64()-task.blockHeader.Number.Uint64() > pruneHeightDiff) {
+					delete(vm.tasks, hash)
+					verifyTaskCounter.Dec(1)
+					verifyTaskFailedMeter.Mark(1)
+					task.Close()
+				}
+			}
+			vm.taskLock.Unlock()
+		case message := <-vm.messageCh:
+			vm.taskLock.RLock()
+			if vt, ok := vm.tasks[message.verifyResult.BlockHash]; ok {
+				vt.messageCh <- message
+			}
+			vm.taskLock.RUnlock()
+		// System stopped
+		case <-vm.bc.quit:
+			vm.taskLock.RLock()
+			for _, task := range vm.tasks {
+				task.Close()
+			}
+			vm.taskLock.RUnlock()
+			return
+		case <-vm.chainHeadSub.Err():
+			return
+		}
+	}
+}
+
+func (vm *remoteVerifyManager) NewBlockVerifyTask(header *types.Header) {
+	for i := 0; header != nil && i <= maxForkHeight; i++ {
+		// if is genesis block, mark it as verified and break.
+		if header.Number.Uint64() == 0 {
+			vm.cacheBlockVerified(header.Hash())
+			break
+		}
+		func(hash common.Hash) {
+			// if verified cache record that this block has been verified, skip.
+			if _, ok := vm.verifiedCache.Get(hash); ok {
+				return
+			}
+			// if there already has a verify task for this block, skip.
+			vm.taskLock.RLock()
+			_, ok := vm.tasks[hash]
+			vm.taskLock.RUnlock()
+			if ok {
+				return
+			}
+
+			if header.TxHash == types.EmptyRootHash {
+				log.Debug("this is an empty block:", "block", hash, "number", header.Number)
+				vm.cacheBlockVerified(hash)
+				return
+			}
+
+			var diffLayer *types.DiffLayer
+			if cached, ok := vm.bc.diffLayerChanCache.Get(hash); ok {
+				diffLayerCh := cached.(chan struct{})
+				<-diffLayerCh
+				vm.bc.diffLayerChanCache.Remove(hash)
```

Comment:
why remove it from `diffLayerChanCache`?

Reply:
> why remove it from `diffLayerChanCache`, we already implement FIFO, don't we?

Yes, we implement FIFO, but I think remove it as soon as we used it is also make sense.

Reply:
Why do you bother to remove it ASAP?

Reply:
I don't understand, when we mean cache, it is a shared cache to be used by many components,  one component does not have the right to purge the item from the cache.

Reply:
> To my understanding, when we mean cache, it is a shared cache to be used by many components, one component does not have the right to purge the item from the cache.

fixed